src/functions/TestResults.ps1

```diff
@@ -352,41 +369,33 @@ function Write-JUnitTestResultAttributes($Result, [System.Xml.XmlWriter] $XmlWri
     $XmlWriter.WriteAttributeString('xmlns', 'xsi', $null, 'http://www.w3.org/2001/XMLSchema-instance')
     $XmlWriter.WriteAttributeString('xsi', 'noNamespaceSchemaLocation', [Xml.Schema.XmlSchema]::InstanceNamespace , 'junit_schema_4.xsd')
     $XmlWriter.WriteAttributeString('name', $Result.Configuration.TestResult.TestSuiteName.Value)
-    $XmlWriter.WriteAttributeString('tests', $Result.PassedCount)
-    $XmlWriter.WriteAttributeString('errors', '0')
+    $XmlWriter.WriteAttributeString('tests', $Result.TotalCount)
+    $XmlWriter.WriteAttributeString('errors', $Result.FailedContainersCount + $Result.FailedBlocksCount)
     $XmlWriter.WriteAttributeString('failures', $Result.FailedCount)
     $XmlWriter.WriteAttributeString('disabled', $Result.NotRunCount + $Result.SkippedCount)
     $XmlWriter.WriteAttributeString('time', ($Result.Duration.TotalSeconds.ToString('0.000', [System.Globalization.CultureInfo]::InvariantCulture)))
 }
 
-function Write-JUnitTestSuiteElements($Node, [System.Xml.XmlWriter] $XmlWriter, [uint16] $Id) {
+function Write-JUnitTestSuiteElements($Container, [System.Xml.XmlWriter] $XmlWriter, [uint16] $Id) {
     $XmlWriter.WriteStartElement('testsuite')
 
-    Write-JUnitTestSuiteAttributes -Action $Node -XmlWriter $XmlWriter -Package $Node.Name -Id $Id
-
-    $testCases = foreach ($al1 in $node.Actions) {
-        if ($al1.Type -ne 'TestCase') {
-            foreach ($al2 in $al1.Actions) {
-                if ($al2.Type -ne 'TestCase') {
-                    foreach ($alt3 in $al2.Actions) {
-                        $path = "$($al1.Name).$($al2.Name).$($alt3.Name)"
-                        $alt3 | & $SafeCommands['Add-Member'] -PassThru -MemberType NoteProperty -Name Path -Value $path
-                    }
-                }
-                else {
-                    $path = "$($al1.Name).$($al2.Name)"
-                    $al2 | & $SafeCommands['Add-Member'] -PassThru -MemberType NoteProperty -Name Path -Value $path
-                }
-            }
-        }
-        else {
-            $path = "$($al1.Name)"
-            $al1 | & $SafeCommands['Add-Member'] -PassThru -MemberType NoteProperty -Name Path -Value $path
-        }
+    if ("File" -eq $Container.Type) {
+        $path = $Container.Item.FullName
+    }
+    elseif ("ScriptBlock" -eq $Container.Type) {
+        $path = "<ScriptBlock>$($Container.Item.File):$($Container.Item.StartPosition.StartLine)"
+    }
+    else {
+        throw "Container type '$($Container.Type)' is not supported."
     }
 
+    Write-JUnitTestSuiteAttributes -Action $Container -XmlWriter $XmlWriter -Package $path -Id $Id
+
+
+    $testCases = [Pester.Factory]::CreateCollection()
+    Fold-Container -Container $Container -OnTest { param ($t) if ($t.ShouldRun) { $testCases.Add($t) } }
```

Comment:
Not significantly, Fold-Container could be optimized to use $null instead of empty scriptblocks for the callbacks that are not provided, but ultimately we are only losing few ms by not using this fix.

If we are after speed then we should iterate in place using just foreaches. But I will wait with that optimization for later when I have perf measuring in place.